File: src/modules/ios/TaskManager.swift
import Foundation

struct BackgroundTask {
    let taskId: String
    let mode: String
    let interval: Int
    let triggers: [String]
    let scheduledTime: Int?
    var isRunning: Bool = false
    var startedAt: Int64? = nil
}

class TaskManager {
    private var tasks: [String: BackgroundTask] = [:]
    private let lock = NSLock()

    func registerTask(_ task: BackgroundTask) -> Bool {
        lock.lock()
        defer { lock.unlock() }

        if tasks[task.taskId] != nil {
            return false
        }
        tasks[task.taskId] = task
        return true
    }

    func unregisterTask(taskId: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }

        return tasks.removeValue(forKey: taskId) != nil
    }

    func getTask(taskId: String) -> BackgroundTask? {
        lock.lock()
        defer { lock.unlock() }

        return tasks[taskId]
    }

    func getAllTasks() -> [BackgroundTask] {
        lock.lock()
        defer { lock.unlock() }

        return Array(tasks.values)
    }

    func getRunningTasks() -> [BackgroundTask] {
        lock.lock()
        defer { lock.unlock() }

        return tasks.values.filter { $0.isRunning }
    }

    func setTaskRunning(taskId: String, running: Bool) {
        lock.lock()
        defer { lock.unlock() }

        if var task = tasks[taskId] {
            task.isRunning = running
            task.startedAt = running ? Int64(Date().timeIntervalSince1970 * 1000) : nil
            tasks[taskId] = task
        }
    }

    func stopAllTasks() {
        lock.lock()
        defer { lock.unlock() }

        for taskId in tasks.keys {
            if var task = tasks[taskId] {
                task.isRunning = false
                task.startedAt = nil
                tasks[taskId] = task
            }
        }
    }

    func hasRunningTasks() -> Bool {
        lock.lock()
        defer { lock.unlock() }

        return tasks.values.contains { $0.isRunning }
    }
}

File: src/modules/android/src/main/java/expo/modules/custombackground/TaskManager.kt
package expo.modules.custombackground

import java.util.concurrent.ConcurrentHashMap

data class BackgroundTask(
    val taskId: String,
    val mode: String,
    val interval: Long,
    val triggers: List<String>,
    val scheduledTime: Long?,
    var isRunning: Boolean = false,
    var startedAt: Long? = null
)

class TaskManager {
    private val tasks = ConcurrentHashMap<String, BackgroundTask>()

    fun registerTask(task: BackgroundTask): Boolean {
        if (tasks.containsKey(task.taskId)) {
            return false
        }
        tasks[task.taskId] = task
        return true
    }

    fun unregisterTask(taskId: String): Boolean {
        return tasks.remove(taskId) != null
    }

    fun getTask(taskId: String): BackgroundTask? {
        return tasks[taskId]
    }

    fun getAllTasks(): List<BackgroundTask> {
        return tasks.values.toList()
    }

    fun getRunningTasks(): List<BackgroundTask> {
        return tasks.values.filter { it.isRunning }
    }

    fun setTaskRunning(taskId: String, running: Boolean) {
        tasks[taskId]?.let { task ->
            tasks[taskId] = task.copy(
                isRunning = running,
                startedAt = if (running) System.currentTimeMillis() else null
            )
        }
    }

    fun stopAllTasks() {
        tasks.keys.forEach { taskId ->
            setTaskRunning(taskId, false)
        }
    }

    fun hasRunningTasks(): Boolean {
        return tasks.values.any { it.isRunning }
    }
}
